[WOR-703] Azure billing project onboarding / modal dialog replacement

cahrens opened the pull request:
New "Wizard" for creating an Azure billing project. This will be shown in the following 2 cases:

1) The user has no billing projects and is an Azure user. 
2) The "Create" button is clicked and an Azure Billing Project is selected.

![image](https://user-images.githubusercontent.com/484484/219786386-96798221-ecdd-43a2-9983-20512734bdc7.png)

Things that probably warrant separate/follow-up tickets:
1) Spinner overlay not filling whole page (I think this  was a pre-existing issue, and unless it is easy to fix we should create a separate ticket and fix it as a follow-up)
2) May want to do some debouncing of validation errors so we don't update them as you are typing (the existing debouncing support seems specific to autocomplete fields and did not work with our inputs, so this could be tricky).

blakery committed:
Blake's commits.

cahrens committed:
Merge issue.

cahrens committed:
Christina's commits.

cahrens committed:
Rename the first version of the wizard.

cahrens committed:
Delete first version of the wizard.

cahrens committed:
Delete CreateAzureBillingProjectModal.

cahrens committed:
Delete unused StepWizard stuff.

cahrens committed:
Clear creating flag (though it doesn't much matter).

cahrens committed:
Remove subscription ID testing hack.

cahrens commented on src/components/input.js:
```diff
@@ -194,16 +194,25 @@ export const NumberInput = forwardRefWithName('NumberInput', ({ onChange, onBlur
+  return createValidatedInput({ inputProps, width, error }, ref)
+})
+
+const createValidatedInput = ({ inputProps, width, error }, ref) => {
```
Created this so that I could put keyboard focus in the subscription ID when the wizard is first rendered. Otherwise the user has to tab through the possibly large number of existing billing projects.

cahrens commented on src/pages/billing/List.js:
```diff
@@ -427,6 +421,19 @@ export const BillingList = ({ queryParams: { selectedName } }) => {
+              search: qs.stringify({ selectedName: billingProjectName, type: 'project' })
+            })
+          }
+        })],
```
Note that I haven't done much testing of _actually_ creating a billing project because of the overhead (and the fact that the billing project cannot be deleted now due to dangling references). Though I did create one with both users and owners and ensured that they showed up on the billing project.

But please someone **do** test create a billing project, ideally with the developer console open so that you can look for any possible errors (I saw a React warning once about state changes after an unmounted component, but then it did not reproduce).

marctalbott replied:
Testing this out and dropping my notes here:
- email validation isn't quite right/isn't very strict. `marc.test` counts as a valid email. When I attempted to create a billing project while adding that user, Rawls returns a 409 saying `Users marc.test have not signed up for Terra` which then manifests in the wizard as an error under the billing project name that says a project by that name already exists. Not sure if the status code should be something else (maybe a 400?) or if we need to handle the error better in the UI.
-  after creating a billing project, it selects the billing project right away and allows you to add members to the billing project even though the project is still creating. I tested it out and it seems to have worked, so maybe it's fine, but it is different from Google where we don't allow you to select a creating billing project. Selecting another project leaves the creating project un-selectable again.

Everything else looked good!

cahrens replied:
The first comment about the 409 should be fixed now that we allow notification of new users. It's true though that the validation is pretty lame, but that's what we get out of the box from the library we are using.

The second comment is something we should consider changing. I made the new billing project selected (which is what we do from the GCP billing wizard, although that is synchronous), but it's true that a billing project in the "creating" state should not be selectable. So perhaps we just go back to the billing page itself with no selected billing project? @blakery thoughts?

blakery replied:
I think it's kind of confusing how it's selected initially, but then if you click away, you can't select it again. 

It might be nice to have it selectable, and then show some indication that it isn't ready yet (greyed-out tabs, a message, etc). But in the scope here, with the current page behavior, it might be better to just return to the main page with no project selected. Although I suppose maybe if you had dozens of these, that could be obnoxious in it's own right, since you would have to go hunting for it.

cahrens replied:
I agree that we should just go back to the billing page itself with no selected project.

cahrens commented on src/pages/billing/NewBillingProjectWizard/AzureBillingProjectWizard/AddUsersStep.ts:
```diff
@@ -0,0 +1,163 @@
+          } else {
+            props.setErrors(null)
+            props.onSetEmails(emails, false)
+            Ajax().Metrics.captureEvent(props.eventKey)
```
Trying to reduce the flurry of events by only sending an event when the email addresses are considered valid.

cahrens commented on src/pages/billing/NewBillingProjectWizard/AzureBillingProjectWizard/AddUsersStep.ts:
```diff
@@ -0,0 +1,163 @@
+            onChange: changed => {
+              props.onAddUsersOrOwners(!changed.target.checked)
+
+              // clear any emails that they previously supplied.
```
Lou requested the clearing behavior.

cahrens commented on src/pages/billing/NewBillingProjectWizard/AzureBillingProjectWizard/AzureBillingProjectWizard.test.ts:
```diff
@@ -0,0 +1,157 @@
+  })
+
+  it('should not fail any accessibility tests in initial state', async () => {
+    expect(await axe(renderResult.container)).toHaveNoViolations()
```
Note that accessibility checks cannot be done on the individual step components because they fail due to `li` being outside of a list.

cahrens commented on src/pages/billing/NewBillingProjectWizard/AzureBillingProjectWizard/AzureBillingProjectWizard.test.ts:
```diff
@@ -0,0 +1,157 @@
+    testStepActive(2)
+    fireEvent.click(getAddUsersRadio())
+    addUserAndOwner('onlyuser@example.com', 'owner@example.com')
+    testStepActive(2) // Have to click in billing project input to become the active step
```
Lou and I discussed this at length-- when you supply email addresses, there is no way to know that the user is "done". Therefore they must move focus to the billing project input to mark the next step as active (and they can't have any email errors).

cahrens commented on src/pages/billing/NewBillingProjectWizard/AzureBillingProjectWizard/AzureBillingProjectWizard.ts:
```diff
@@ -0,0 +1,160 @@
+
+  const [isBusy, setIsBusy] = useState(false)
+
+  const createBillingProject = _.flow(
```
A brought much of this code over from `CreateAzureBillingProjectModal`. I gave up on using `useLoadedState` because I wasn't sure how to handle error conditions (and I'm just not familiar with it yet).

blakery replied:
Better done than perfect.
But you can just render errors when the state of the loaded data is `'Error'`:
```typescript
const [data, updateData] = useLoadedData(updateDataMethod)

return .... 
  // render error
  data.status === 'Error' &&  // error component 

```

cahrens replied:
I will eventually get used to `useLoadedData`, but at this point I'm not a huge fan. One of the things that I find confusing is that it's less obvious if all the ajax calls have error handling around them.

cahrens commented on src/pages/billing/NewBillingProjectWizard/AzureBillingProjectWizard/AzureBillingProjectWizard.ts:
```diff
@@ -0,0 +1,160 @@
+    setManagedApp(managedApp)
+  }
+
+  const step1HasNoErrors = !!subscriptionId && !!managedApp
```
This step transition logic isn't overly tested-- just the happy path. I ran out of time to figure out how to write more thorough tests.

cahrens commented on src/pages/billing/NewBillingProjectWizard/AzureBillingProjectWizard/AzureBillingProjectWizard.ts:
```diff
@@ -0,0 +1,160 @@
+      createReady: step1HasNoErrors && step2HasNoErrors && !!billingProjectName && !projectNameErrors && !isBusy
+    })
+  ]),
+  isBusy && spinnerOverlay]) // TODO: overlay problem: not filling the whole screen.
```
I'm pretty sure this was a pre-existing issue, but it's very noticeable with this long view. It would be great to fix it, but probably as a follow-up ticket.

blakery replied:
Probably pre-existing. 
It's going to be kind of obnoxious to fix, when just managing the state in the individual components, instead of having some sort of state store.
I remember Miguel saying he had started to implement such things in places, but I don't remember the details.

cahrens commented on src/pages/billing/NewBillingProjectWizard/AzureBillingProjectWizard/CreateNamedProjectStep.test.ts:
```diff
@@ -0,0 +1,115 @@
+
+    // Assert
+    verifyEnabled(getBillingProjectInput())
+    // @ts-ignore
```
Perhaps you can figure out how to avoid this?

cahrens commented on src/pages/billing/NewBillingProjectWizard/AzureBillingProjectWizard/AzureSubscriptionStep.ts:
```diff
@@ -0,0 +1,163 @@
+  }
+}, apps)
+
+// @ts-ignore
```
Perhaps you can figure out how to avoid these ignores?

blakery replied:
This is primarily because the [validation library we're using](https://github.com/ansman/validate.js) is outdated and unmaintained (note the archive notice on the page), so the types are outdated.

There are a couple of options for this:
1. Switch to a maintained library
2. Create a type definition for the library in the types folder of the project
3. cast to any (in practice, casting to any is pretty similar to adding ts-ignore)

cahrens commented on src/pages/billing/NewBillingProjectWizard/StepWizard/StepFields.ts:
```diff
@@ -41,17 +43,6 @@ export const StepInfo = ({ children, style }: StepFieldsProps) => div({
   style: { ...primaryStepTextStyle as CSSProperties, ...style }
 }, [children])
 
-export const StepFieldForm = ({ children, style }: StepFieldsProps) => form({
```
Unused.

cahrens commented on src/pages/billing/NewBillingProjectWizard/StepWizard/StepHeader.ts:
```diff
@@ -1,13 +1,12 @@
-  h3({ style: { fontSize: 18, marginTop: 0, marginRight: '1rem' } }, [title]),
-  span({ style: { fontSize: '1rem', lineHeight: '22px', width: '75%' } }, [children]),
+export const StepHeader = ({ title, style }: StepHeaderProps) => header({ style: { width: '100%', display: 'flex', flexDirection: 'row', ...style } }, [
+  h3({ style: { fontSize: 18, marginTop: 0, marginRight: '1rem' } }, [title])
```
We went with the style of always having the text below the Step X header.

cahrens committed:
Remove extra event call.

cahrens committed:
Move all eventing out of the step components.

marctalbott commented on src/pages/billing/NewBillingProjectWizard/AzureBillingProjectWizard/AddUsersStep.ts:
```diff
@@ -0,0 +1,153 @@
+        disabled: props.disabled,
+        onChange: emails => {
+          const individualEmails = getIndividualEmails(emails)
+          const emailErrors = validateEmails(individualEmails)
```
Could we validate when focus leaves the step or otherwise avoid displaying validation errors while a user is in the process of typing out emails?

cahrens replied:
@blakery added debouncing, and it feels natural to me.

aherbst-broad replied:
+1 on validating once focus leaves; this is validating on each keystroke and is overkill imo. What's the lift to make this not fire so often?

cahrens replied:
@aherbst-broad you must not have the most recent version of the code-- check that you have the last commit.

cahrens replied:
I don't think we want to wait until focus leaves the field because then you don't get the feedback until you move to work on something else. The delayed validation felt good to me.

marctalbott commented on src/libs/ajax/Billing.js:
```diff
@@ -26,10 +26,11 @@ export const Billing = signal => ({
     return await fetchRawls('billing/v2',
       _.mergeAll([authOpts(), jsonBody(
-        { projectName, managedAppCoordinates: { tenantId, subscriptionId, managedResourceGroupId } }
+        { projectName, managedAppCoordinates: { tenantId, subscriptionId, managedResourceGroupId }, members }
```
I was under the impression that we want users to be able to invite others to Terra while creating a billing project, in which case we'd need to set `inviteUsersNotFound` to true or Rawls will throw if a user is invited during billing project creation

cahrens replied:
Yes, we should set inviteUsersNotFound to true. That will likely also take care of the 409 issue.

blakery committed:
add a spinner variant to allow for overriding styling

blakery committed:
use custom styles to make spinner overlay fill page

blakery committed:
clean up props/state for subscriptionId and managed apps and switch to useLoadedData hook

blakery committed:
remove incorrect dependency

setManagedApps was added because of a linter warning, but it isn't a dependency in the normal sense, and causes infinite re-renders

blakery committed:
debounce subscription id input

blakery committed:
add parameter to billing project creation to invite missing users to terra

blakery committed:
debounce email inputs

blakery committed:
fix tests broken by debouncing input

blakery commented on src/pages/billing/NewBillingProjectWizard/AzureBillingProjectWizard/AzureSubscriptionStep.test.ts:
```diff
@@ -0,0 +1,176 @@
+  }
+
+  it('shows the spinner overlay while the call to list managed apps is in progress', async () => {
+    renderAzureSubscriptionStep({})
```
The noManagedAppsTestCase function was testing that the setIsBusy function in the props was called.
Since the spinner was fixed, we no longer need that function, so I had to add a separate test case in order to check the page was showing the spinner while the call was in progress.

blakery commented on src/pages/billing/NewBillingProjectWizard/AzureBillingProjectWizard/AzureSubscriptionStep.test.ts:
```diff
@@ -0,0 +1,176 @@
+  let onManagedAppSelectedEvent
+  let renderResult
+
+  const renderAzureSubscriptionStep = props => {
```
I thought I was going to need to set some props separately for some tests. It turned out it wasn't necessary, but I figured I'd leave the ability anyway.

cahrens replied:
Ha, I was just coming to the PR to comment that this method wasn't necessary and the test could be simplified by doing the setup in `beforeEach`. But I don't feel that strongly about it.

blakery commented on src/pages/billing/NewBillingProjectWizard/AzureBillingProjectWizard/AzureSubscriptionStep.ts:
```diff
@@ -0,0 +1,158 @@
+      } else {
+        console.error(state.error)
+      }
+      setSubscriptionIdError(h(NoManagedApps))
```
we could display a more detailed message here, possibly derived from the response...

cahrens replied:
We actually don't want a more detailed response. As we discussed on some recent tickets, we don't want to differentiate if the user doesn't have access to the subscription vs. if they do but have no managed apps.

In the code being deleted, we presented a different message if they had access but there were no managed apps.

blakery commented on src/pages/billing/NewBillingProjectWizard/AzureBillingProjectWizard/AddUsersStep.ts:
```diff
@@ -0,0 +1,167 @@
+  const [debouncedEmails, setDebouncedEmails] = useState<ReactNode>()
+  const [debouncedErrors, setDebouncedErrors] = useState<ReactNode>()
+
+  useEffect(() => {
```
I think we could make a debounced version of the `ValidatedInput` input, for wider consumption

blakery committed:
changed spinner overlay position to 'fixed' to fill entire page

cahrens commented on src/pages/billing/NewBillingProjectWizard/AzureBillingProjectWizard/AzureSubscriptionStep.ts:
```diff
@@ -0,0 +1,158 @@
+        })
+      }
+    }, inputDebounce)
+    return () => clearTimeout(timeoutId)
```
@blakery let's remove the debouncing on this field to simplify the code. It will also remove the need to deal with `setManagedApps` as a dependency.

Users are most likely going to paste into this field, but even if they type they won't see the error message changing with every keystroke as the UUID will be invalid up until they type the last digit.

cahrens replied:
Ugh, looks like `setManagedApps` dependency issue was introduced by the code changes for `useLoadedData`. I don't like ignoring the warning-- @msilva-broad is there another option to avoid having to ignore dependencies?

https://github.com/DataBiosphere/terra-ui/pull/3790/files/03a438e91d59e905b89df07ad687948e8a3b9844#diff-55e8cfc1bf896fdf1117687f5a3aa9a8c761f88cc34c289c45ecba66f3f694aeR100

cahrens replied:
@msilva-broad we no longer need the `useEffect` callback, so the need to disable the eslint warning is moot.

But I'm still curious about the general issue-- is there a way to update React/eslint so that it treats `useLoadedData` as a dependency that it wan't warn about (in the same way that it doesn't warn about `useState`)?

blakery committed:
removed debounce from subscription id step

blakery committed:
fixed email input debounce default value

aherbst-broad commented on src/pages/billing/List.js:
```diff
@@ -306,6 +306,8 @@ export const BillingList = ({ queryParams: { selectedName } }) => {
     if (type === cloudProviders.azure) {
       setCreatingBillingProject(type)
+      // Show the Azure wizard instead of the selected billing project.
+      Nav.history.replace({ search: '' })
```
I'm not sure what this is doing -- it looks like it's mucking with the browser history and setting the query string to nothing, but how does that end up showing the azure wizard?

cahrens replied:
If we leave the search string on the current selected billing project, we wouldn't show the create billing project "wizard" in its place. The search string specifies the selected billing project.

cahrens commented:
@blakery I have looked through your code changes, and they LGTM. I didn't test actually creating a billing project though (and with users to invite)… I'm saving that for when we think we have all the changes in place because then I get stuck with another undeletable billing project.

These are the 2 things I think still need to be decided before merging--

1. Are we OK with the current level of eventing, or do we wish to be more selective/drop some of the events?
2. Do we want to return to the billing page with no selected project after billing project creation? I think we probably do because we don't allow billing projects in the "creating" state to be selected normally. I think the friendliest option would be to have a different view of "creating" projects that at least allowed you to select them and have a message about their state so that it's obvious what happened (as opposed to having to look for the billing project name on the left and deduce that it can't be selected because it's creating), but that's likely out-of-scope here.

aherbst-broad commented on src/pages/billing/NewBillingProjectWizard/AzureBillingProjectWizard/AzureSubscriptionStep.test.ts:
```diff
@@ -0,0 +1,175 @@
+    expect(captureEvent).not.toHaveBeenCalled()
+  })
+
+  const noManagedAppsTestCase = async listAzureManagedApplications => {
```
👍 I like this pattern of extracting a common setup + assertion that can be reused across test cases.

aherbst-broad commented on src/pages/billing/NewBillingProjectWizard/AzureBillingProjectWizard/AzureSubscriptionStep.ts:
```diff
@@ -0,0 +1,148 @@
+              id: subscriptionInputId,
+              placeholder: 'Azure Subscription ID',
+              onChange: subscriptionIdChanged,
+              value: subscriptionId ?? '',
```
TIL about the TS `??` operator.

blakery replied:
It returns the right hand side of the operator, unless it evaluates as false-y, in which case it returns the right hand side instead. It's the equivalent of:
`subscriptionId ? subscriptionId : ''`

aherbst-broad reviewed (commented):
Looking good. We spoke in slack but I'll note here: I'm seeing a ton of calls to bard during my keystrokes in various text boxes, it'd be good to reduce the amount of API chattiness there.

cahrens committed:
Limit eventing, don't select created project.

aherbst-broad reviewed (approved):
🚢

cahrens committed:
Fix react warning and a spinner.

blakery commented on src/pages/billing/NewBillingProjectWizard/AzureBillingProjectWizard/AzureBillingProjectWizard.ts:
```diff
@@ -0,0 +1,173 @@
+  })
+
+  useEffect(() => {
+    let isMounted = true // Necessary to avoid React warning about updates in unmounted component after project created.
```
If we update react to v18 (🤞), the warning will go away without this

cahrens replied:
Good to know!
